tools/isobuild/import-scanner.js

```diff
@@ -0,0 +1,431 @@
+import assert from "assert";
+import {isString, has, keys, each, without} from "underscore";
+import {sha1} from "../fs/watch.js";
+import {matches as archMatches} from "../utils/archinfo.js";
+import {findImportedModuleIdentifiers} from "./js-analyze.js";
+import {
+  pathJoin,
+  pathRelative,
+  pathNormalize,
+  pathDirname,
+  pathBasename,
+  pathExtname,
+  statOrNull,
+  readFile,
+  convertToPosixPath,
+} from "../fs/files.js";
+
+export default class ImportScanner {
+  constructor({
+    name,
+    bundleArch,
+    sourceRoot,
+    extensions = [".js", ".json"],
+    usedPackageNames = {},
+    nodeModulesPath,
+  }) {
+    assert.ok(isString(sourceRoot));
+
+    this.name = name;
+    this.bundleArch = bundleArch;
+    this.sourceRoot = sourceRoot;
+    this.usedPackageNames = usedPackageNames;
+    this.nodeModulesPath = nodeModulesPath;
+    this.absPathToOutputIndex = {};
+    this.outputFiles = [];
+    this.extensions = extensions;
+  }
+
+  addInputFiles(...args) {
+    const Ap = Array.prototype;
+    const flatFiles = Ap.concat.apply(Ap, args);
```

Comment:
Oh wait, no, I remember now: I was trying to support both of the following styles:
* passing a single array of files to `addInputFiles`
* passing multiple file arguments to `addInputFiles`

I guess it's a sign of something that I forgot why this code was worthwhile… :blush: